File: services/api/apps/documents/visibilidad.py
"""Que se estaba viendo cuando se abrio la observacion: `F4.7`, leido sin creerle nada.

**Es la mitad del punto de vista que faltaba.** `F4.1` cerro la camara —desde donde se miraba— y el
BCF seguia saliendo con `DefaultVisibility="true"` y las excepciones vacias, o sea **el modelo
entero a la vista**. Cuando el hallazgo se encontro aislando una planta o apagando la disciplina de
arquitectura, eso no es un detalle que falte: es una afirmacion falsa, y quien abre el archivo en
Solibri ve el edificio completo con el problema tapado por lo que precisamente se habia apagado.

**Y llega desde el visor, asi que es dato hostil**, exactamente igual que la camara. La regla de que
lado se escribe vive en `bim-core` —`ladoDeVisibilidad` y `visibilidadBcf`, con sus pruebas— y aca
**solo se valida**: repetir la regla daria dos implementaciones de lo mismo, que es la forma segura
de que se separen. Lo que se comprueba es la forma del dato, porque lo que se guarda sale despues en
un archivo que se manda al mandante.

**Una visibilidad mala no es un error del que avisar.** Se descarta y la observacion se guarda sin
ella: el BCF vuelve a salir con el modelo entero, que es lo que hacia antes de que esto existiera.
"""

import json

#: Tope de excepciones, **el mismo que `MAXIMO_EXCEPCIONES` de `bim-core`**.
#:
#: No es un limite del formato: BCF no pone ninguno. Es el punto a partir del cual el archivo deja
#: de ser util —cinco mil `<Component>` ya son mas de cien kilobytes de XML por observacion— y
#: ademas acota lo que puede entrar en un `JSONField` desde fuera.
MAXIMO_EXCEPCIONES = 5_000

#: Tope del JSON que se acepta, en caracteres.
#:
#: Cada GUID son 22 caracteres mas comillas y coma: veinticinco por excepcion. Con holgura para las
#: llaves y el resto de la estructura, el tope es lo que ocupa el maximo de excepciones y nada mas.
LARGO_MAXIMO = MAXIMO_EXCEPCIONES * 26 + 200

#: Largo de un GUID de IFC comprimido.
LARGO_GUID = 22

#: El alfabeto de base64 de IFC. El mismo que comprueba `ObservacionesDeRevisionAPI` para el ancla.
_EXTRA_GUID = "_$"


def _es_guid(valor) -> bool:
    """La misma forma que se le exige al ancla: 22 caracteres del alfabeto de IFC.

    **Se comprueba uno por uno y no se confia en el visor.** Un GUID mal formado no rompe nada
    visible hoy: produce un `<Component IfcGuid="...">` que el otro extremo ignora en silencio, y
    entonces el viewpoint muestra algo distinto de lo que se anoto.
    """
    return (
        isinstance(valor, str)
        and len(valor) == LARGO_GUID
        and all(caracter.isalnum() or caracter in _EXTRA_GUID for caracter in valor)
    )


def leer(crudo) -> dict:
    """La visibilidad que venga del visor, validada. `{}` si no hay o si no sirve.

    Devuelve un diccionario listo para guardar en `Observacion.visibilidad`, con las mismas claves
    que escribe `bim-core` y que lee `bcf.py`. **Un diccionario vacio significa «sin restriccion»**,
    que es un estado normal y no un fallo: el viewpoint sale con el modelo entero.

    Acepta la cadena JSON que manda el visor o el diccionario ya decodificado, porque los dos
    caminos existen —la tarjeta de nota manda JSON en el cuerpo, y una llamada interna puede pasar
    el objeto—.
    """
    if crudo is None:
        return {}

    if isinstance(crudo, str):
        if not crudo or len(crudo) > LARGO_MAXIMO:
            return {}
        try:
            datos = json.loads(crudo)
        except (ValueError, TypeError):
            return {}
    else:
        datos = crudo

    if not isinstance(datos, dict):
        return {}

    por_defecto = datos.get("porDefecto")
    # `isinstance(1, bool)` es falso pero `isinstance(True, int)` es verdadero: aca se exige el bool
    # de verdad, porque un `1` que se cuela como `True` invierte el sentido del viewpoint entero.
    if not isinstance(por_defecto, bool):
        return {}

    excepciones = datos.get("excepciones")
    if not isinstance(excepciones, list) or len(excepciones) > MAXIMO_EXCEPCIONES:
        return {}

    # Se filtran los que no son GUID y se quitan los repetidos, conservando el orden: dos modelos
    # abiertos pueden traer el mismo elemento y la lista se duplicaria sin decir nada nuevo.
    limpias = []
    vistos = set()
    for candidato in excepciones:
        if not _es_guid(candidato) or candidato in vistos:
            continue
        vistos.add(candidato)
        limpias.append(candidato)

    # **Sin excepciones utilizables no se guarda nada**, y por el lado `false` eso importa de
    # verdad: `DefaultVisibility="false"` con la lista vacia es un viewpoint que **apaga el modelo
    # entero** y se abre en negro. Antes que escribir eso, no se escribe visibilidad.
    if not limpias:
        return {}

    return {"porDefecto": por_defecto, "excepciones": limpias}

File: packages/bim-core/src/registro/visibilidad.ts
/**
 * Qué se estaba viendo cuando se abrió la observación: `F4.7`.
 *
 * **Es la mitad del punto de vista que faltaba.** `F4.1` cerró la cámara —desde dónde se miraba— y
 * el BCF salía diciendo `DefaultVisibility="true"` con las excepciones vacías, o sea **el modelo
 * entero a la vista**. Cuando el hallazgo se encontró aislando una planta o apagando la disciplina
 * de arquitectura, eso no es un detalle que falte: es una afirmación falsa. Quien abre el archivo
 * en Solibri ve el edificio completo y el problema tapado por lo que precisamente se había apagado.
 *
 * ## Los dos lados de `Visibility`, y por qué hay que elegir
 *
 * BCF no guarda "lo que se ve": guarda un **valor por defecto y sus excepciones**.
 *
 * | `DefaultVisibility` | Qué significa                                              |
 * | ------------------- | --------------------------------------------------------- |
 * | `true`              | Se ve todo **menos** los componentes de `Exceptions`       |
 * | `false`             | No se ve nada **salvo** los componentes de `Exceptions`    |
 *
 * Los dos pueden describir la misma pantalla, y lo que cambia es **cuántos componentes hay que
 * escribir**. Apagando tres vigas de un modelo de veinte mil elementos, el lado `true` escribe tres
 * líneas y el lado `false` escribiría 19 997. Aislando una planta es al revés. Así que la regla es
 * escribir **el lado corto**, que además de producir un archivo manejable es lo que hace que el
 * viewpoint diga lo mismo con menos.
 *
 * ## Y hay un tope, porque un viewpoint enorme no lo abre nadie
 *
 * Con los dos lados por encima del tope no se escribe visibilidad: se vuelve al comportamiento de
 * antes —el modelo entero— y **eso es honesto**, porque un viewpoint que ningún visor termina de
 * leer no informa de nada. El caso solo aparece con medio modelo apagado a mano, que es raro; el
 * aislamiento, que es lo frecuente, deja el lado corto en unas decenas.
 *
 * El GUID de IFC es la identidad —regla de `AGENTS.md`— y por eso lo que sale de aquí son GUID y
 * nunca `localId`: los identificadores del motor cambian entre versiones del modelo y entre
 * herramientas, y un viewpoint que apunte a ellos queda huérfano en la siguiente exportación.
 */

import { isIfcGuid } from "../identity/ifcGuid.js";

/** La visibilidad de un viewpoint de BCF, con sus dos campos y nada más. */
export interface VisibilidadBcf {
  /** `DefaultVisibility`: `true` se ve todo menos las excepciones; `false`, al revés. */
  readonly porDefecto: boolean;
  /** Los GUID de las excepciones, sin repetidos y en orden estable. */
  readonly excepciones: readonly string[];
}

/** De qué lado se escriben las excepciones. */
export type LadoDeVisibilidad = "ocultos" | "visibles";

/**
 * Cuántas excepciones se aceptan en un viewpoint.
 *
 * **No es un límite del formato**: BCF no pone ninguno. Es el punto a partir del cual el archivo
 * deja de ser útil — cinco mil `<Component>` ya son más de cien kilobytes de XML por observación, y
 * un BCF con treinta temas así empieza a pesar más que el modelo del que habla.
 *
 * Cubre de sobra lo que se hace en la práctica: aislar deja el lado corto en unas decenas, y apagar
 * una disciplina de un modelo federado, en unos miles. Lo que deja fuera es medio modelo apagado a
 * mano, que es raro y para el que callarse es mejor que escribir un archivo que nadie abre.
 *
 * **El mismo número está en `services/api/apps/documents/visibilidad.py`**, que es quien lo hace
 * cumplir de verdad: lo que llega al servidor lo escribe cualquiera.
 */
export const MAXIMO_EXCEPCIONES = 5_000;

/**
 * Qué lado sale más corto, o `null` si no hay nada que decir o no cabe.
 *
 * Recibe **cuentas y no listas** a propósito: resolver un GUID por elemento cuesta una consulta al
 * modelo, y con esto quien llama resuelve **solo el lado que va a escribir**. En un modelo grande
 * con tres elementos apagados, la diferencia es tres búsquedas contra veinte mil.
 *
 * Devuelve `null` en dos casos, y ninguno es un error:
 *
 * 1. **No hay nada oculto**: no hay ninguna restricción que afirmar.
 * 2. **Los dos lados pasan del tope**: ver el docstring del módulo.
 */
export function ladoDeVisibilidad(ocultos: number, visibles: number): LadoDeVisibilidad | null {
  if (!Number.isFinite(ocultos) || !Number.isFinite(visibles)) return null;
  if (ocultos <= 0) return null;

  // Empatados se elige `ocultos`, que deja `DefaultVisibility="true"` — el valor que ya escribía el
  // exportador y el que cualquier lector interpreta sin sorpresas.
  const lado: LadoDeVisibilidad = ocultos <= visibles ? "ocultos" : "visibles";
  const cuantos = lado === "ocultos" ? ocultos : visibles;
  return cuantos > MAXIMO_EXCEPCIONES ? null : lado;
}

/**
 * La visibilidad lista para el viewpoint, o `null` si lo que llegó no sirve.
 *
 * **Filtra los GUID que no lo son y quita los repetidos.** Lo primero porque un elemento sin GUID
 * válido no se puede nombrar en un BCF —el atributo `IfcGuid` tiene una forma, y escribir cualquier
 * cosa produce un archivo que el otro extremo ignora o rechaza—; lo segundo porque dos modelos
 * abiertos pueden traer el mismo elemento y la lista se duplicaría sin decir nada nuevo.
 *
 * Devuelve `null` cuando no queda ninguna excepción utilizable: con el lado `ocultos` eso significa
 * que no hay nada que ocultar, y con el lado `visibles` significaría **un viewpoint que apaga el
 * modelo entero**, que es lo peor que se puede escribir — se prefiere no decir nada.
 */
export function visibilidadBcf(
  lado: LadoDeVisibilidad,
  guids: readonly (string | null | undefined)[],
): VisibilidadBcf | null {
  const limpios: string[] = [];
  const vistos = new Set<string>();
  for (const guid of guids) {
    if (typeof guid !== "string" || !isIfcGuid(guid) || vistos.has(guid)) continue;
    vistos.add(guid);
    limpios.push(guid);
  }

  if (limpios.length === 0 || limpios.length > MAXIMO_EXCEPCIONES) return null;
  return { porDefecto: lado === "ocultos", excepciones: limpios };
}

/**
 * `true` si este GUID se ve, según la visibilidad guardada.
 *
 * Es la lectura del mismo dato en el otro sentido, y la necesita el visor para **volver a dejar la
 * pantalla como estaba** al abrir la observación. Vive acá y no en el visor porque la regla —qué
 * significa cada lado de `DefaultVisibility`— es la misma que se usó para escribirlo, y tenerla dos
 * veces es la forma segura de que las dos se separen.
 */
export function seVe(visibilidad: VisibilidadBcf, guid: string): boolean {
  const esExcepcion = visibilidad.excepciones.includes(guid);
  return esExcepcion ? !visibilidad.porDefecto : visibilidad.porDefecto;
}
